Add support for more subnets than AZs

michaelmoussa opened the pull request:
**Replaces #9 with upstream PR so that CircleCI build will run**

---

> Currently, the module does not support having a larger list of public or private subnets than AZs. For example if you only have two AZ but six private subnets, the module will only build two private subnets.
> 
> This PR should solve the issue.

daynelucasrax committed:
Add support for more subnets than AZs

johnctitus reviewed (changes requested):
I am not quite sure what we are trying to solve with this change.  I understand that the subnets are limited to the number of AZs, but that is actually by intent.  

My concern with this change, is we end up with a nonstandard configuration.  For instance this would allow a 3 AZ deployment to consist of 2 public subnets and 7 private subnets.  In fact, using our default values, this would deploy a 2 AZ solution with 3 public and 3 private subnets.  Anything we change, we should ensure subnets are deployed evenly and consistently across all selected AZs.  I am sure there will be special cases where we do not want something deployed in every AZ, but that's the exception, and should be handled outside the module, at least in my opinion. 

I think it will also be somewhat confusing if all the subnets are just numeric iterations (what is the difference between `MyVPC-PrivateSubnet1` and `MyVPC-PrivateSubnet3`)  If we proceed with this change, I would like to enable setting a custom name for the type of subnet.

michaelmoussa commented:
@johnctitus the use case is a 2 AZ VPC that requires 1 public subnet in each AZ (2 total) and 3 private subnets in each AZ (6 total). If one supplies 6 private subnet CIDR ranges to the module, it will only create 2 of them since `count` is limited to `2` because of the AZs.

While the following isn't specifically the motivation behind this particular change, this is also a common pattern in PCI environments that auditors still look for.

As for concerns about inconsistency like having fewer subnets than AZs or having some AZs with more subnets than others, I'm not sure if that's something we can effectively mitigate against with the tools available to us in the current version of HCL. The only thing that comes to mind is modifying the `count` value with an expression that evaluates to zero if `<num subnets> % <num AZs>` doesn't evaluate to zero, but I'm not sure if that'll cause Terraform to complain about the "count cannot be computed in modules" nonsense.

Finally, for custom naming, how about an optional `(public|private)_subnet_names` parameter that is a `list` of `string` values? For each iteration, if `${var.(public|private)_subnet_names[count.index]}` exists, we set the `Name` tag to whatever that value is set to? Otherwise, use the current default values (and same deal for the route table associations?)

johnctitus commented:
@michaelmoussa Thanks for the background.  If we proceed with the change, I think I would like to adjust how we are going about it.  

Instead I propose a `private_subnets_per_az` variable, and we could set our resource counts to `count = "${var.az_count * var.private_subnets_per_az}"` .  This would ensure consistent application across all AZs since it will always be a multiple of the AZ count.  It will be destructive if the AZ counts change, but this is already the case.  We could handle the custom naming with a `private_subnet_names` variable defaulting to `["Private"]` and referenced by `element(var.private_subnet_names, count.index / var.az_count)`.  We have done something very similar with the EKS module
https://github.com/rackspace-infrastructure-automation/aws-terraform-eks/blob/master/main.tf#L64-L69

Would of course duplicate these variables for public subnets, if we really think its necessary.  I personally think 1 public is sufficient, but as long as we are consistent, the more the merrier.

michaelmoussa commented:
Thanks for the feedback @johnctitus. That sounds good to me as well. I'll update the PR with that as soon as I can.

michaelmoussa committed:
Determine subnets per AZ via `_subnets_per_az` variable

* Updated default values for private and public cidr ranges for
compatibility (cannot evenly divide 3 subnets across 2 AZs)
* Modified tests, examples, and README accordingly

michaelmoussa committed:
Allow customization of subnet names

michaelmoussa commented:
@johnctitus first revision attempt done. What do you think?

**Specific things to note:**

1. I modified the default `xxxxxxx_cidr_ranges` lists to have length 2 instead of 3 because the previous value didn't make sense when mixed with the new `xxxxxxx_subnets_per_az` count variable and a default `az_count` of 2.
    * Actually, the original defaults _did_ run without error, but wouldn't create all of the listed subnets because it'd stop after the 2nd due to the `count` being based on `az_count`.
2. I updated the NATGW resource counts accordingly in order to keep the "NAT Gateways will be created in all public subnets" bullet point from the README accurate. Is that a behavior we'd want to keep, though? If I'm not mistaken, we need one NATGW _per-AZ_, not per public subnet.
    * Should we instead create one NATGW for each AZ's "first" public subnet in the event that there is more than 1? It _does_ seem like an extreme edge case to me, but I felt I should call it out anyway for completeness' sake.
3. I'm not 100% happy with the naming convention for AZs just yet. Consider an example with 2 private subnets per AZ with `private_subnet_names = ["App", "Data"]`. With the current logic, they'd be named as follows: `MyVpc-App1`, `MyVpc-App2`, `MyVpc-Data3`, `MyVpc-Data4`. Now imagine you've never seen this environment before and you're looking at the subnet list. "Hey, why are there 4 Data subnets and only 2 App subnets? Oh, wait... where are "Data1" and "Data2?" Are there actually only 2, or are 2 of them missing?"
    * What do you think of using the AZ a subnet resides in as the suffix, rather than a simple incrementing value? So instead of the 4 generated names shown above, we'd instead have: `MyVpc-App-a`, `MyVpc-App-b`, `MyVpc-Data-a`, `MyVpc-Data-b`. Now, at a glance, we know (a) the VPC they live in, (b) their overall purpose, and (c) which AZ they live in.
    * In the case of multiple subnets that would have the same name, we could tack on an integer suffix like: `MyVpc-Private-a1`, `MyVpc-Private-a2`, etc., to indicate "1st private subnet in AZ A, 2nd private subnet in AZ A" and so on.

The additional proposed changes in (2) and (3) shouldn't take much effort to do, but I wanted to ask first before taking the time.

johnctitus commented on variables.tf:
```diff
@@ -82,21 +82,63 @@ variable "public_cidr_ranges" {
   default = [
     "172.18.168.0/22",
     "172.18.172.0/22",
-    "172.18.176.0/22",
```
I would prefer we leave the third CIDR.  This was actually to maintain equivalence to our CloudFormation template, and provide default values for a standard 3 AZ deployment.

johnctitus commented on variables.tf:
```diff
@@ -82,21 +82,63 @@ variable "public_cidr_ranges" {
   default = [
     "172.18.0.0/21",
     "172.18.8.0/21",
-    "172.18.16.0/21",
```
I would prefer we leave the third CIDR.  This was actually to maintain equivalence to our CloudFormation template, and provide default values for a standard 3 AZ deployment.

johnctitus commented on main.tf:
```diff
@@ -117,13 +116,13 @@ resource aws_route "private_routes" {
 
 resource aws_route_table_association "private_route_association" {
-  count          = "${var.az_count}"
+  count          = "${length(var.private_cidr_ranges)}"
```
This line should probably be since we need one association per subnet, the counts should match.
```suggestion
  count          = "${var.az_count * var.private_subnets_per_az}"
```

johnctitus commented on main.tf:
```diff
@@ -117,13 +116,13 @@ resource aws_route "private_routes" {
 
 resource aws_route_table_association "public_route_association" {
-  count          = "${var.az_count}"
+  count          = "${length(var.public_cidr_ranges)}"
```
This line should probably be since we need one association per subnet, the counts should match.
```suggestion
  count          = "${var.az_count * var.public_subnets_per_az}"
```

michaelmoussa replied:
@johnctitus I believe `var.az_count * var.public_subnets_per_az` _should_ always `= length(var.public_cidr_ranges)` anyway, but this would account for the _technically_ valid (albeit weird) case of `<`. Will update.

johnctitus commented on main.tf:
```diff
@@ -70,22 +70,21 @@ resource aws_nat_gateway "nat" {
   map_public_ip_on_launch = false
-
-  tags = "${merge(local.base_tags, map("Name", format("%s-PrivateSubnet%d", var.vpc_name, count.index + 1)), var.custom_tags)}"
+  tags                    = "${merge(local.base_tags, map("Name", format("%s-%s%d", var.vpc_name, element(var.private_subnet_names, count.index / var.az_count), count.index + 1)), var.custom_tags)}"
```
I think to fix the issue with subnet naming we could update as follows
```suggestion
  format("%s-%s%d", var.vpc_name, element(var.private_subnet_names, count.index / var.az_count), (count.index % var.az_count) + 1))"
```

I would also like to see if we can break up this line.  Its getting a bit long, and if we could break up into multiple lines would help readability.

johnctitus commented on main.tf:
```diff
@@ -70,22 +70,21 @@ resource aws_nat_gateway "nat" {
   availability_zone       = "${element(local.azs, count.index)}"
   map_public_ip_on_launch = true
-  tags                    = "${merge(local.base_tags, map("Name", format("%s-PublicSubnet%d", var.vpc_name, count.index + 1)), var.custom_tags)}"
+  tags                    = "${merge(local.base_tags, map("Name", format("%s-%s%d", var.vpc_name, element(var.public_subnet_names, count.index / var.az_count), count.index + 1)), var.custom_tags)}"
```
I think to fix the issue with subnet naming we could update as follows
```suggestion
  format("%s-%s%d", var.vpc_name, element(var.private_subnet_names, count.index / var.az_count), (count.index % var.az_count) + 1))"
```

I would also like to see if we can break up this line.  Its getting a bit long, and if we could break up into multiple lines would help readability.

michaelmoussa replied:
~The `diff` in the "Suggested change" looks a little weird (I think it'd delete the `tags =` portion, so I'll apply the change by hand to be safe.~

Actually, I'm curious. Will test and rebase to clean up any mess.

![image](https://user-images.githubusercontent.com/183833/47368589-ed150a00-d6af-11e8-9c6d-1eb1f71713ed.png)

johnctitus commented on main.tf:
```diff
@@ -58,7 +58,7 @@ resource aws_eip "nat_eip" {
 
 resource aws_nat_gateway "nat" {
-  count         = "${var.build_nat_gateways ? var.az_count : 0}"
+  count         = "${var.build_nat_gateways ? (var.az_count * var.public_subnets_per_az) : 0}"
```
I think if we leave this line as is we accomplish the 1 NAT gateway per AZ, which I believe is the best way to proceed.

michaelmoussa replied:
OK - I'll update the docs then to indicate one per AZ rather than one in each public subnet.

johnctitus reviewed (changes requested):
>I modified the default xxxxxxx_cidr_ranges lists to have length 2 instead of 3 because the previous value didn't make sense when mixed with the new xxxxxxx_subnets_per_az count variable and a default az_count of 2.

As noted in the code review, I would prefer to leave the CIDR defaults as is, if not add additional CIDRs.  The original intent was to mirror CloudFormation which provided defaults for up to 3 AZs.  With this change, I could see us adding defaults to support 6, 9, or even more default CIDRs, with the point being require a minimum amount of data input when instantiating the module.

> I updated the NATGW resource counts accordingly in order to keep the "NAT Gateways will be created in all public subnets" bullet point from the README accurate. Is that a behavior we'd want to keep, though? If I'm not mistaken, we need one NATGW per-AZ, not per public subnet.
> - Should we instead create one NATGW for each AZ's "first" public subnet in the event that there is more than 1? It does seem like an extreme edge case to me, but I felt I should call it out anyway for completeness' sake.

I would say we only create one NAT gateway per AZ.  I can't think of any reason why we would need multiple NAT Gateways per AZ, and the increased cost would be a good reason to avoid it.  For those times where it is needed, I think it would be better to handle outside of the module itself.

> I'm not 100% happy with the naming convention for AZs just yet. Consider an example with 2 private subnets per AZ with private_subnet_names = ["App", "Data"]. With the current logic, they'd be named as follows: MyVpc-App1, MyVpc-App2, MyVpc-Data3, MyVpc-Data4. Now imagine you've never seen this environment before and you're looking at the subnet list. "Hey, why are there 4 Data subnets and only 2 App subnets? Oh, wait... where are "Data1" and "Data2?" Are there actually only 2, or are 2 of them missing?"

Agreed.  I proposed an update that I think would be an improvement, but I think either of your methods would be fine as well.  Important thing is to make it easy as possible to determine at a glance where and what things are.

michaelmoussa committed:
One NATGW per AZ, not per public subnet

michaelmoussa committed:
Change route table association calculation

Base on az_count * # subnets per AZ rather than total # of
public/private cidr ranges. *Should* result in the same value in most
cases, but this variation accounts for the edge case where az_count * #
subnets is less than count of ranges.

michaelmoussa committed:
Increment subnet name suffix per name, not overall

(e.g. Foo1, Foo2, Bar1, Bar2 versus Foo1, Foo2, Bar3, Bar4)

michaelmoussa committed:
Restore 3rd default CIDR pub/priv range; update README

michaelmoussa committed:
Improve subnet naming expression readability (depending on who you ask)

johnctitus commented on variables.tf:
```diff
@@ -86,6 +86,28 @@ variable "public_cidr_ranges" {
+EOF
+
+  type    = "list"
+  default = ["Private"]
```
OK, a few more and then I promise I will stop.  Looks like we swapped the defaults for the public and private subnet name params.  We should update the default and descriptions for this one to `Public`

johnctitus commented on variables.tf:
```diff
@@ -97,6 +119,28 @@ variable "private_cidr_ranges" {
+EOF
+
+  type    = "list"
+  default = ["Public"]
```
And we should change default and description to use `Private`

johnctitus reviewed (changes requested):
OK, one more change requested to correct defaults\descriptions for the subnet name parameters.  With apologies for missing the first go round.  Otherwise, looks good.

michaelmoussa committed:
Fix swapped public/private naming

michaelmoussa commented:
@johnctitus should be good to go now, CircleCI willing. :)

johnctitus reviewed (approved):
👍

twistedgrim reviewed (approved):
👍
